File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter04/WhatSizeWithBindingConverter/WhatSizeWithBindingConverter/FormattedStringConverter.cpp
#include "pch.h"            // includes sstream!
#include "FormattedStringConverter.h"

using namespace WhatSizeWithBindingConverter;

using namespace std;
using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Interop;

Object^ FormattedStringConverter::Convert(Object^ value, TypeName targetType, Object^ parameter, String^)
{
    if (targetType.Name == "String" && parameter != nullptr)
    {
        int decimals = _wtoi(dynamic_cast<String^>(parameter)->Data());
        wostringstream strStream;
        strStream.imbue(locale(""));
        strStream.precision(decimals);
        strStream << (double)value;
        return ref new String(strStream.str().c_str());
    }
    return value;
}

Object^ FormattedStringConverter::ConvertBack(Object^ value, TypeName targetType, Object^ parameter, String^)
{
    return value;
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter04/WhatSizeWithBindingConverter/WhatSizeWithBindingConverter/FormattedStringConverter.cs
using System;
using System.Globalization;
using Windows.UI.Xaml.Data;

namespace WhatSizeWithBindingConverter
{
    public class FormattedStringConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, string language)
        {
            if (value is IFormattable && 
                parameter is string && 
                !String.IsNullOrEmpty(parameter as string) &&
                targetType == typeof(string))
            {
                if (String.IsNullOrEmpty(language))
                    return (value as IFormattable).ToString(parameter as string, null);

                return (value as IFormattable).ToString(parameter as string, 
                                                        new CultureInfo(language));
            }

            return value;
        }
        public object ConvertBack(object value, Type targetType, object parameter, string language)
        {
            return value;
        }
    }
}
